File: part-3/backend/models/visitor.go
package models

import (
	"time"

	"github.com/jinzhu/gorm"
)

// Visitor struct
type Visitor struct {
	gorm.Model
	IpAddress string    `gorm:"unique_index;column:ip_address" json:"ip_address"`
	Location  string    `gorm:"column:location" json:"location"`
	Timezone  string    `gorm:"column:timezone" json:"timezone"`
	LastVisit time.Time `gorm:"index;column:last_visit" json:"last_visit"`
	Visits    uint      `gorm:"index;column:visits" json:"visits"`
}

File: part-3/frontend/src/app/core/models/visitor.ts
// Province : tỉnh / thành phố
export interface Visitor {
  ip_address: string;
  location: string;
  timezone: string;
  last_visit: string;
  visits: number;
}

export interface GetOneVisitorResponse {
  data: Visitor;
  error: string;
}

export interface Get100VisitorsResponse {
  data: Visitor[];
  error: string;
}